### ISSUE TITLE
[HOLD E/PR #535] [$250] Chat - In mWeb, text is centered but in Android text is dropping down

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#52827]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** V9. 0.64-3
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Y
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Open both mWeb and app
2. Open a chat
3. Enter > # hai

## Expected Result:
In mWeb, text is centered but in Android text must not drop down.

## Actual Result:
In mWeb, text is centered but in Android text is dropping down.

## Workaround:
Unknown
## Platforms:

- [x] Android: Standalone
- [x] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/d37b1dac-6824-423c-bc2a-bdf2b921ef3e

 </details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021863815952930925048</li>
        <li>Upwork Job ID: 1863815952930925048</li>
        <li>Last Price Increase: 2024-12-10</li>
    </ul>
</details>

**Comment by QichenZhu** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Headings are not vertically centered on Android.

### What is the root cause of that problem?
The line height is enlarged on Android to fix [overlapping lines](https://github.com/Expensify/react-native-live-markdown/issues/61).

<img width="839" alt="Screenshot 2024-11-23 at 1 27 58 AM" src="https://github.com/user-attachments/assets/950f60ef-3070-4304-9276-6c1db78f084b">

<p></p>

However, the text is bottom-aligned in [`MarkdownLineHeightSpan`](https://github.com/Expensify/react-native-live-markdown/blob/cec2feae7463398b8d377a231be8671e1698e4a0/android/src/main/java/com/expensify/livemarkdown/spans/MarkdownLineHeightSpan.java#L13-L17).

<img width="986" alt="Screenshot 2024-11-23 at 1 40 05 AM" src="https://github.com/user-attachments/assets/818a620a-04bf-4e67-b915-197f036d0e54">

### What changes do you think we should make in order to solve the problem?
Fine-tune the `fm.ascent`, `fm.descent`, `fm.top`, `fm.bottom` properties to vertically center the text. The code below refers to [this implementation](https://github.com/facebook/react-native/blob/d424bb9d7cf101dc5609e8f787ba8af2a33d0262/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/internal/span/CustomLineHeightSpan.kt#L15-L55). It’s not the final version and might need confirmation from the design team.

```Java
  @Override
  public void chooseHeight(CharSequence text, int start, int end, int spanstartv, int lineHeight, Paint.FontMetricsInt fm) {
    float leading = mLineHeight - ((-fm.ascent) + fm.descent);
    fm.ascent -= (int)Math.ceil(leading / 2.0);
    fm.descent += (int)Math.floor(leading / 2.0);

    // The top of the first line, and the bottom of the last line, may influence bounds of the
    // paragraph, so we match them to the text ascent/descent. It is otherwise desirable to allow
    // line boxes to overlap (to allow too large glyphs to be drawn outside them), so we do not
    // adjust the top/bottom of interior line-boxes.
    if (start == 0) {
      fm.top = fm.ascent;
    }
    if (end == text.length()) {
      fm.bottom = fm.descent;
    }
  }
```

#### Screenshots

| Before | After |
| -- | -- |
| <img alt="Before" src="https://github.com/user-attachments/assets/87c00640-6e68-4dec-b95c-9f64e59ba0cf"> | <img alt="After" src="https://github.com/user-attachments/assets/d21a6a31-ff97-440f-a435-b154b472b5f8"> |

### What alternative solutions did you explore? (Optional)
N/A

**Comment by alexpensify** [comment]
I didn't get to test this one before going OOO. This week, I'll try to log in to confirm the next steps.

**Comment by alexpensify** [comment]
No update

**Comment by alexpensify** [comment]
@alitoshmatov - I was able to replicate the experience. Can you please verify if this proposal will fix the issue? Thanks!

**Comment by alitoshmatov** [comment]
@QichenZhu Are you suggestion that this should be fixed in `react-native-live-markdown`

**Comment by QichenZhu** [comment]
@alitoshmatov That’s correct.

**Comment by alitoshmatov** [comment]
@tomekzaw Can you check out this issue where heading in android is pushed down noticeably. And we have [this proposal](https://github.com/Expensify/App/issues/52827#issuecomment-2493694585) which suggests to update height calculation for android. 

I suspect this issue is related to https://github.com/Expensify/react-native-live-markdown/issues/522

What do you think

**Comment by tomekzaw** [comment]
@alitoshmatov Indeed, this should be fixed on Android with https://github.com/Expensify/react-native-live-markdown/pull/535 telling from the screenshots in https://github.com/Expensify/react-native-live-markdown/pull/535#issuecomment-2525081847.

**Comment by alitoshmatov** [comment]
Great. I guess we can hold this issue then. 

@alexpensify Can you update the issue to hold for https://github.com/Expensify/react-native-live-markdown/pull/535

**Comment by alexpensify** [comment]
Done

**Comment by alexpensify** [comment]
## Heads up, I will be offline until Wednesday, December 18, 2024, and will not actively watch over this GitHub during that period.

If this GitHub requires an urgent update, please ask for help in the #expensify-open-source Slack Room. If the inquiry can wait, I'll review it when I return online.

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/react-native-live-markdown/pull/535 is still under review

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/react-native-live-markdown/pull/535 is still under review

**Comment by alexpensify** [comment]
Weekly Update: Same as last week

**Comment by alexpensify** [comment]
Weekly Update: Waiting on https://github.com/Expensify/react-native-live-markdown/pull/535

**Comment by alexpensify** [comment]
Weekly Update: https://github.com/Expensify/react-native-live-markdown/pull/535 is still under review

**Comment by alexpensify** [comment]
Weekly update: same as last week

**Comment by alexpensify** [comment]
Weekly Update: No movement in the PR

**Comment by tomekzaw** [comment]
Just FYI, I wasn't actively working on this PR because I was hoping that RN 0.76 upgrade would fix this. Unfortunately this is not the case so I'll need to resume working on this PR once I complete my current tasks.

**Comment by alexpensify** [comment]
Oh no, thank you for the update. Keep us posted when you have a better timeline to get to this one. Thanks!

**Comment by alexpensify** [comment]
Weekly Update: Waiting for @tomekzaw to have a free cycle

**Comment by alexpensify** [comment]
Weekly Update: Same

**Comment by alexpensify** [comment]
Weekly Update: Waiting for @tomekzaw

--- 

## 🚨 Heads up! I'll be offline until Wednesday, March 12, 2025, and won’t be actively monitoring this GitHub during that time. 

If this GitHub requires an urgent update, please reapply the `bug` label or reach out in the `#expensify-open-source` Slack room for help. Otherwise, I'll pick things up when I'm back online. Thanks!  

--- 

Update: The proposal is under review

**Comment by alexpensify** [comment]
Weekly Update: Same update, but I'm back online.

**Comment by tomekzaw** [comment]
Update: I'm working on it, Android already works fine, iOS is almost working but needs some final touches. Latest updates here:
* https://github.com/Expensify/react-native-live-markdown/pull/535#issuecomment-2724999165
* https://github.com/Expensify/react-native-live-markdown/pull/535#issuecomment-2725023102
* https://github.com/Expensify/react-native-live-markdown/pull/535#issuecomment-2725057396
* https://github.com/Expensify/react-native-live-markdown/pull/535#issuecomment-2725087982

**Comment by alexpensify** [comment]
Thank you, @tomekzaw, for the update!

---  

🚨 Heads up! I'll be offline until Monday, April 7, 2025, and won’t be actively monitoring this GitHub during that time. 

If this GitHub requires an urgent update, please reapply the `bug` label or reach out in the `#expensify-open-source` Slack room for help. Otherwise, I'll pick things up when I'm back online. Thanks!

**Comment by alexpensify** [comment]
@tomekzaw is there any update here? Thanks!

**Comment by tomekzaw** [comment]
@alexpensify I've pushed some changes to https://github.com/Expensify/react-native-live-markdown/pull/535 but found out that a proper solution requires some small changes in react-native code. There's also another PR (https://github.com/Expensify/react-native-live-markdown/pull/520) that I'd like to merge prior to this one.

**Comment by alexpensify** [comment]
Still waiting on https://github.com/Expensify/react-native-live-markdown/pull/535

**Comment by alexpensify** [comment]
Update: Still waiting on the issue above

**Comment by alexpensify** [comment]
See above

**Comment by alexpensify** [comment]
Weekly Update: Still waiting on the related PR

**Comment by alexpensify** [comment]
Weekly Update: Hold

**Comment by alexpensify** [comment]
Weekly Update: On Hold

**Comment by brianlee-expensify** [comment]
still on hold

**Comment by brianlee-expensify** [comment]
followed up in PR [above](https://github.com/Expensify/App/issues/52827#issuecomment-3615335388)

**Comment by war-in** [comment]
Hi! I'm taking over the issue from @tomekzaw and would like to confirm the expected behaviour. Those are recordings from the current production releases


Web:

https://github.com/user-attachments/assets/087bc6d3-3291-47a1-bfd5-b6beb5cd36da

iOS:

https://github.com/user-attachments/assets/e8de56b5-057d-4903-b19c-09ad748be5ee

Android:

https://github.com/user-attachments/assets/4775bbd4-a854-4eff-99ae-6f1e7c392293

Looks like the original issue is gone and all platforms behave similarly. Should we still adjust something or are we fine with it?

**Comment by brianlee-expensify** [comment]
@war-in if the original issue is gone, then i think we can close out - thanks! closing out now
